qa/tasks/cephfs/test_volumes.py

```diff
@@ -2376,6 +2377,69 @@ def test_subvolume_clear_earmark(self):
         self.assertEqual(get_earmark, "")
+    
+    def test_earmark_on_non_existing_subvolume(self):
+        subvolume = "non_existing_subvol"
+        earmark = "nfs.test"
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolume", "earmark", "set", self.volname, subvolume, earmark)
+
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolume", "earmark", "get", self.volname, subvolume)
+
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolume", "earmark", "rm", self.volname, subvolume)
+    
+    def test_get_remove_earmark_when_not_set(self):
+        # Create a subvolume without setting an earmark
+        subvolume = self._gen_subvol_name()
+        self._fs_cmd("subvolume", "create", self.volname, subvolume)
+
+        # Attempt to get an earmark when it's not set
+        get_earmark = self._fs_cmd("subvolume", "earmark", "get", self.volname, subvolume)
+        self.assertEqual(get_earmark, "")
+
+        # Attempt to remove an earmark when it's not set
+        self._fs_cmd("subvolume", "earmark", "rm", self.volname, subvolume)
+    
+    def test_set_invalid_earmark(self):
+        # Create a subvolume
+        subvolume = self._gen_subvol_name()
+        self._fs_cmd("subvolume", "create", self.volname, subvolume)
+
+        # Attempt to set an invalid earmark
+        invalid_earmark = "invalid_format"
+        expected_message = (
+            f"Invalid earmark specified: '{invalid_earmark}'. A valid earmark should "
+            "either be empty or start with 'nfs' or 'smb', followed by dot-separated "
+            "non-empty components."
+        )
+        try:
+            self._fs_cmd("subvolume", "earmark", "set", self.volname, subvolume, invalid_earmark)
+        except CommandFailedError as ce:
+            self.assertIn(expected_message, ce.output)
+    
+    def test_earmark_on_deleted_subvolume_with_retained_snapshot(self):
+        subvolume = self._gen_subvol_name()
+        snapshot = self._gen_subvol_snap_name()
+
+        # Create subvolume and snapshot
+        self._fs_cmd("subvolume", "create", self.volname, subvolume)
+        self._fs_cmd("subvolume", "snapshot", "create", self.volname, subvolume, snapshot)
+
+        # Delete subvolume while retaining the snapshot
+        self._fs_cmd("subvolume", "rm", self.volname, subvolume, "--retain-snapshots")
+
+        # Define the expected error message
+        error_message = f'subvolume "{subvolume}" is removed and has only snapshots retained'
+
+        # Test cases for setting, getting, and removing earmarks
+        for operation in ["set", "get", "rm"]:
+            try:
+                self._fs_cmd("subvolume", "earmark", operation, self.volname, subvolume, "smb" if operation == "set" else "")
+            except CommandFailedError as ce:
+                self.assertEqual(ce.exitstatus, errno.ENOENT)
```

Comment:
I think this would fail with ENOTSUP  ? Did you validate this locally ?

Reply:
Yes! I did check locally 
```
[ceph: root@ceph-node-00 /]# ceph fs subvolume earmark set fs1 sv1 --earmark smb
Error ENOENT: subvolume "sv1" is removed and has only snapshots retained

```

Reply:
I ran this command for removing subvol
```
ceph fs subvolume rm fs1 sv1 --retain-snapshots
```

Reply:
Yes, right! It's ENOENT